proposals/file-types.md

```diff
@@ -0,0 +1,150 @@
+
+## Detailed design
+[design]: #detailed-design
+
+- We add the `file` modifier to the [class](https://github.com/dotnet/csharpstandard/blob/draft-v6/standard/classes.md#1422-class-modifiers), [struct](https://github.com/dotnet/csharpstandard/blob/draft-v6/standard/structs.md#1522-struct-modifiers), [interface](https://github.com/dotnet/csharpstandard/blob/draft-v6/standard/interfaces.md#1722-interface-modifiers), and [enum](https://github.com/dotnet/csharpstandard/blob/draft-v6/standard/enums.md#183-enum-modifiers) modifier sets in the grammar.
+- The `file` modifier can only be used on a top-level type.
+
+### Accessibility
+No accessibility modifiers can be used in combination with `file` on a type. `file` is treated as an independent concept from accessibility. Since file types can't be nested, only the default accessibility `internal` is usable with `file` types.
+
+```cs
+public file class C1 { } // error
+internal file class C2 { } // error
+file class C3 { } // ok
+```
+
+### Naming
+The implementation guarantees that `file` types in different files with the same name will be distinct to the runtime. The type's accessibility and name in metadata is implementation-defined. The intention is to permit the compiler to adopt any future access-limitation features in the runtime which are suited to the feature. It's expected that in the initial implementation, an `internal` accessibility would be used and an unspeakable generated name will be used which depends on the file the type is declared in.
+
+### Lookup
+We amend the [member lookup](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/expressions.md#115-member-lookup) section as follows (new text in **bold**):
+
+> - Next, if `K` is zero, all nested types whose declarations include type parameters are removed. If `K` is not zero, all members with a different number of type parameters are removed. When `K` is zero, methods having type parameters are not removed, since the type inference process ([§11.6.3](expressions.md#1163-type-inference)) might be able to infer the type arguments.
+> - **Next, let *F* be the compilation unit which contains the expression where member lookup is occurring. All members which are file types and are not declared in *F* are removed from the set.**
+> - **Next, if the set of accessible members contains file types, all non-file types are removed from the set.**
+
+#### Remarks
+These rules disallow usage of file types outside the file in which they are declared.
+
+These rules also permit *shadowing* of a non-file type by a file type:
+```cs
+// File1.cs
+class C
+{
+    public static void M() { }
+}
+```
+
+```cs
+// File2.cs
+file class C
+{
+    public static void M() { }
+}
+
+class Program
+{
+    static void Main()
+    {
+        C.M(); // refers to the 'C' in File2.cs
+    }
+}
+```
+
+### Attributes
```

Comment:
This makes sense, but a concern is that one main scenario for attributes is to be checked via reflection. For file types, that doesn't seem practical. 
I'm wondering whether we should disallow this (using file-types as attributes). Do we have a motivating scenario?